### PR TITLE
Refactor embeddings

### PR DESCRIPTION
**Pull request by enyst**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR proposes a refactoring of embeddings to:
- start to separate the lower level backend (embeddings, torch) from long term memory class
- add batch processing of events
- clean up the code from outdated retries (which don't work well with background threads)
- separate functions for easier profiling.

I did some profiling on cpu and metal for local embeddings, and updated the default num_workers to 3. All of this is currently not in use, disabled by config, until we get it better.

This PR is a small step towards offering some backend for experiments here, it will hopefully become less difficult to test stuff than it is when one needs to rewrite too much at once.

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by enyst** [comment]
Added unit tests, but they're excluded from CI, because we have CI installing poetry --without llama-index. I'm thinking that maybe it's not very important right now to have them, because installing with it implies llama and torch. That will add time and memory use to the workflows. But as we move forward we will have to add them at some point.

File: openhands/memory/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        else:
+            from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+
+            # initialize the local embedding model
+            local_embed_model = HuggingFaceEmbedding(
+                model_name='BAAI/bge-small-en-v1.5'
+            )
+
+            # for local embeddings, we need torch
+            import torch
```
Comment (xingyaoww) [inline openhands/memory/embeddings.py:118]: (not for this one, but for future) I'm actually thinking if we can add dynamic import for torch - so torch is only required if we use local embed model 🤔

**Review by xingyaoww — APPROVED** [review]
LGTM! Thanks for adding this -- this will prepare us better for embedding-based context management next!

Instead of context condensation we experimented earlier, we could probably just do a simple approach:
1. Add ALL events to embedding store
2. Only keep X number of most recent events (or with a threshold of Y tokens) in context window
3. Do a some sort of RAG:
  - Given initial instruction, most recent context, retrieve the relevant events from vector DB.
  - Optionally, filter events by similarity (e.g., if top 3 has similar probability than the rest - probably means no important stuff retrieved)
  - Send that to an LLM to condense / summarize key info related to current step

Now that we've got a better evaluation system - hopefully we will be able to develop / iterate on this faster! I actually hope this will be the next major thing we will be looking at after the editing PR.

File: openhands/memory/embeddings.py
Comment (xingyaoww) [inline openhands/memory/embeddings.py:1]: And actually, what do you feel if we put this under `openhands/utils`? Bc i imagine we could use this embeddings later in other places (e.g., inside the runtime to embed the whole codebase)

File: openhands/memory/embeddings.py (hunk 1 above)
Comment (enyst) [reply openhands/memory/embeddings.py:118]: Definitely! Will take another look.

**Comment by enyst** [comment]
That sounds good, absolutely! I have a branch with an experimental memory-enabled agent, I think I'm going to rebuild it from scratch. We could use it to see a little more clearly the options, and figure out what else we need in the backend to try various approaches.

File: openhands/memory/embeddings.py
Comment (enyst) [reply openhands/memory/embeddings.py:1]: Done

**Pull request merged**